pinot-transport/src/main/java/com/linkedin/pinot/routing/HelixExternalViewBasedRouting.java

```diff
@@ -176,49 +195,171 @@ public void markDataResourceOnline(String tableName, ExternalView externalView,
     if (externalView == null) {
       return;
     }
+
+    // Check if there are relevant changes to the state or if we can skip this update
+    if (!isRoutingTableRebuildRequired(tableName, externalView, instanceConfigList)) {
+      return;
+    }
+
+    // Build the routing table
+    buildRoutingTable(tableName, externalView, instanceConfigList);
+  }
+
+  private boolean isRoutingTableRebuildRequired(String tableName, ExternalView externalView,
+      List<InstanceConfig> instanceConfigs) {
+    // In unit tests, always rebuild the routing table
+    if (_helixManager == null) {
+      return true;
+    }
+
+    // Do we know about this table?
+    if (!_lastKnownExternalViewVersionMap.containsKey(tableName)) {
+      LOGGER.info("Routing table for table {} requires rebuild due to it being newly added", tableName);
+      return true;
+    }
+
+    // Check if the znode version changed
     int externalViewRecordVersion = externalView.getRecord().getVersion();
-    if (_routingTableLastKnownZkVersionMap.containsKey(tableName)) {
-      long lastKnownZkVersion = _routingTableLastKnownZkVersionMap.get(tableName);
-      if (externalViewRecordVersion == lastKnownZkVersion) {
-        LOGGER.info(
-            "No change on routing table version (current version {}, last known version {}), do nothing for table {}",
-            externalViewRecordVersion, lastKnownZkVersion, tableName);
-        return;
+    int lastKnownExternalViewVersion = _lastKnownExternalViewVersionMap.get(tableName);
+
+    if (externalViewRecordVersion != lastKnownExternalViewVersion) {
+      LOGGER.info(
+          "Routing table for table {} requires rebuild due to external view change (current version {}, last known version {})",
+          tableName, externalViewRecordVersion, lastKnownExternalViewVersion);
+      return true;
+    }
+
+    // Check if there are relevant instance config changes
+    Map<String, InstanceConfig> lastKnownInstanceConfigs = _lastKnownInstanceConfigsForTable.get(tableName);
+    if (lastKnownInstanceConfigs == null || lastKnownInstanceConfigs.isEmpty()) {
+      LOGGER.info("Routing table for table {} requires rebuild due to empty/null previous instance configs", tableName);
+      return true;
+    }
+
+    // Gather relevant incoming instance configs
+    Map<String, InstanceConfig> currentRelevantInstanceConfigs = new HashMap<>();
+    for (InstanceConfig incomingInstanceConfig : instanceConfigs) {
+      String instanceName = incomingInstanceConfig.getInstanceName();
+
+      if (lastKnownInstanceConfigs.containsKey(instanceName)) {
```

Comment:
We don't care because our routing table does not contain this new instance. If a new instance is added to serve this table, we will get an external view change, at which point we'll keep track of this new instance's configuration.